app/controllers/DeleteSchemeController.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2018 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import javax.inject.Inject
+
+import config.FrontendAppConfig
+import connectors.DataCacheConnector
+import controllers.actions._
+import forms.DeleteSchemeFormProvider
+import play.api.data.Form
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent}
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import utils.Navigator
+import views.html.deleteScheme
+
+import scala.concurrent.Future
+
+class DeleteSchemeController @Inject() (
+                                                     appConfig: FrontendAppConfig,
+                                                     override val messagesApi: MessagesApi,
+                                                     dataCacheConnector: DataCacheConnector,
+                                                     navigator: Navigator,
+                                                     authenticate: AuthAction,
+                                                     getData: DataRetrievalAction,
+                                                     requireData: DataRequiredAction,
+                                                     formProvider: DeleteSchemeFormProvider
+                                                   ) extends FrontendController with I18nSupport with Retrievals {
+
+  private val form: Form[Boolean] = formProvider()
+
+  def onPageLoad: Action[AnyContent] = (authenticate andThen getData andThen requireData).async {
+    implicit request =>
+      retrieveSchemeName { schemeName =>
+      Future.successful(Ok(deleteScheme(appConfig, form, schemeName)))
+    }
+  }
+
+  def onSubmit: Action[AnyContent] = (authenticate andThen getData andThen requireData).async {
+    implicit request =>
+      retrieveSchemeName { schemeName =>
+        form.bindFromRequest().fold(
+          (formWithErrors: Form[_]) =>
+            Future.successful(BadRequest(deleteScheme(appConfig, formWithErrors, schemeName))),
+          {
+            case true =>
+              dataCacheConnector.removeAll(request.externalId)
```

Comment:
At the moment this fires the removeAll but doesn't wait to see if successful. If we map on this if the removeall is successful then we can redirect to the What you will need, if it fails then the map will not be called and the future will not be successful, but at least this will mean that user doesn't proceed with data they have asked to be cleared and we will have visibility of the error in the logs.